lessons/advanced/protocols.md

```diff
@@ -0,0 +1,78 @@
+---
+layout: page
+title: Protocols
+category: advanced
+order: 12
+lang: en
+---
+
+In this lesson we are going to look at Protocols, and how to leverage them in Elixir.
+
+{% include toc.html %}
+
+## What Are Protocols
+Protocols are the chosen method for polymorphism in Elixir.
+One of the primary pains of Erlang is that extending a given API for a newly defined type is quite difficult.
+What Elixir does to avoid this pain is dispatch function calls dynamically based on the type of a value provided.
+This typically is not an issue in Object Oriented languages, but the workaround is necessary in our environment here.
+If a data type implements a given protocol, it will have it available to dispatch to.
+Let's demonstrate this with a simple example looking at `to_string/1`.
+
+```elixir
+iex> to_string(5)
+"5"
+iex> to_string(12.4)
+"12.4"
+iex> to_string("foo")
+"foo"
+```
+
+As you can see, we have called this on multiple types.
+We've demonstrated that the function works on integers, floats, and strings as well.
+If we try to do it on a tuple (or anything else that has not implemented it), we get a different result:
+
+```elixir
+to_string({:foo})
+** (Protocol.UndefinedError) protocol String.Chars not implemented for {:foo}
+    (elixir) lib/string/chars.ex:3: String.Chars.impl_for!/1
+    (elixir) lib/string/chars.ex:17: String.Chars.to_string/1
+```
+
+As you can see, we get a protocol error.
+This is because there is no implementation for the Protocol in the tuples.
+To implement a protocol, `defimpl/3` is our tool of choice.
+
+## Working With Protocols
```

Comment:
How about we use an example that can be applied to a few different types?  Something like:

We know how to implement a protocol but how we do we define a new one?  For our example we'll implement `to_atom/1`, let's see how to do that with `defprotocol`:

```elixir
defprotocol AsAtom do
  def to_atom(data)
end

defimpl AsAtom, for: Atom do
  def to_atom(atom), do: atom
end

defimpl AsAtom, for: BitString do
  defdelegate to_atom(string), to: String
end

defimpl AsAtom, for: List do
  defdelegate to_atom(list), to: List
end
```

Here we've defined our protocol and it's expected method, `to_atom/1`, along with implementing it for a few types.  Now that we have our protocol, let's put it to use in IEx:

```elixir
iex> import AsAtom
AsAtom
iex> to_atom("string")
:string
iex> to_atom(:an_atom)
:an_atom
iex> to_atom([1, 2])
:"\x01\x02"
```